src/CloudNative.CloudEvents/CloudEvent.cs

```diff
@@ -1,123 +1,202 @@
+        public object this[CloudEventAttribute attribute]
         {
-            this.attributes = attributes;
-            var extensionMap = new Dictionary<Type, ICloudEventExtension>();
-            if (extensions != null)
+            // TODO: Should we validate that the known attribute is compatible with the requested one?
+            // For example, if we know about a CloudEventAttribute with a type of "string", and instead
+            // one is requested with the same name but a type of "timestamp", the value returned may be
+            // unhelpful...
+            get => attributeValues.GetValueOrDefault(Preconditions.CheckNotNull(attribute, nameof(attribute)).Name);
+            set
             {
-                foreach (var extension in extensions)
+                Preconditions.CheckNotNull(attribute, nameof(attribute));
+                string name = attribute.Name;
+                var knownAttribute = GetAttribute(name); 
+
+                // TODO: Are we happy to add the extension in even if the value is null?
+                if (knownAttribute is null)
+                {
+                    if (!attribute.IsExtension)
+                    {
+                        throw new ArgumentException("Cannot add an unknown non-extension attribute to an event.");
+                    }
+                    extensionAttributes[name] = attribute;
+                }
+
+                if (value is null)
                 {
-                    extensionMap.Add(extension.GetType(), extension);
-                    extension.Attach(this);
+                    attributeValues.Remove(name);
+                    return;
                 }
+                // TODO: We could convert the attribute value here instead? Or is that a bit too much "magic"?
+                attributeValues[name] = attribute.Validate(value);
             }
-            Extensions = new ReadOnlyDictionary<Type, ICloudEventExtension>(extensionMap);
         }
 
         /// <summary>
-        /// CloudEvent 'datacontenttype' attribute. Content type of the 'data' attribute value.
-        /// This attribute enables the data attribute to carry any type of content, whereby
-        /// format and encoding might differ from that of the chosen event format.
+        /// Sets or fetches the value associated with the given attribute name.
+        /// Setting a value of null removes the value from the event, if it exists.
+        /// If the attribute is not known in this event, fetching the value always returns null, and
+        /// setting the value add a new extension attribute with the given name, and a type of string.
+        /// (The value for an unknown attribute must be a string or null.)
         /// </summary>
-        /// <see cref="https://github.com/cloudevents/spec/blob/master/spec.md#contenttype"/>
-        public ContentType DataContentType
+        public object this[string attributeName]
```

Comment:
It seems like this may not be very discoverable to users. Have you considered a dictionary property or even having CloudEvent implement IDictionary?

Reply:
(Again, I missed this before - apologies. I really wish GitHub had a simple "go to next comment" option that would include diffs that it wasn't showing by default.)

Implementing IDictionary could be tricky - it's not obvious what the Count would be, for example. We've also got the issue that we've got two indexers - one for string and one for CloudEventAttribute, and we probably don't want to implement different generic versions of IDictionary.

I'm happy to consider this again later - I'll add a TODO at the top of the file.